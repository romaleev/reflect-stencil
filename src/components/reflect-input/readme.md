# reflect-input



<!-- Auto Generated Below -->


## Events

| Event          | Description | Type                     |
| -------------- | ----------- | ------------------------ |
| `inputChanged` |             | `CustomEvent<InputData>` |


## Methods

### `clear() => Promise<void>`

Clears input contents

#### Returns

Type: `Promise<void>`



### `getData() => Promise<InputData>`



#### Returns

Type: `Promise<InputData>`



### `setData(data: InputData) => Promise<void>`



#### Returns

Type: `Promise<void>`




## Dependencies

### Used by

 - [reflect-form](../reflect-form)

### Graph
```mermaid
graph TD;
  reflect-form --> reflect-input
  style reflect-input fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
