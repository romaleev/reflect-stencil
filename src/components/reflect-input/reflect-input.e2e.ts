import { newE2EPage } from '@stencil/core/testing'

const MockValidInput = { name: 'Roma', email: 'email@example.com' }
const MockInvalidInput = { name: 'R1', email: 'email@examplecom' }
const Red = 'rgb(255, 0, 0)'

describe('reflect-input', () => {
  it('renders', async () => {
    const page = await newE2EPage()
    await page.setContent('<reflect-input></reflect-input>')

    const element = await page.find('reflect-input')
    expect(element).toHaveClass('hydrated')
  });

  it('renders Name field correctly', async () => {
    const page = await newE2EPage()
    await page.setContent('<reflect-input></reflect-input>')
    const inputName = await page.find('reflect-input >>> input[name="name"]')

    await inputName.type(MockValidInput.name)
    await page.waitForChanges()

    expect((await inputName.getComputedStyle()).color).not.toEqual(Red)

    await inputName.type(MockInvalidInput.name)
    await page.waitForChanges()

    await page.waitFor(500) // wait for red color fade in
    expect((await inputName.getComputedStyle()).color).toEqual(Red)
  })

  it('renders Email field correctly', async () => {
    const page = await newE2EPage()
    await page.setContent('<reflect-input></reflect-input>')
    const inputEmail = await page.find('reflect-input >>> input[name="email"]')

    await inputEmail.type(MockValidInput.email)
    await page.waitForChanges()

    expect((await inputEmail.getComputedStyle()).color).not.toEqual(Red)

    await inputEmail.type(MockInvalidInput.email)
    await page.waitFor(500) // wait for red color fade in

    expect((await inputEmail.getComputedStyle()).color).toEqual(Red)
  })
})
