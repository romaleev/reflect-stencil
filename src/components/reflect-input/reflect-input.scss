$reflect-input-color: #19197C;
$reflect-input-background-color: #FFF;
$reflect-input-border-color: lighten($reflect-input-color, 60%);
$reflect-input-focus-color: #FD6400;
$font: 1rem Arial, Helvetica, sans-serif;

:host {
  display: block;

  input, textarea {
    transition: all 0.30s ease-in-out;
    outline: none;
    box-sizing: border-box;
    width: 100%;
    background: $reflect-input-background-color;
    margin-bottom: 1rem;
    border: 1px solid $reflect-input-border-color;
    padding: 0.75rem;
    color: $reflect-input-color;
    font: $font;

    &:invalid {
      color: red;
    }

    &:focus {
      box-shadow: 0 0 5px $reflect-input-focus-color;
      border: 1px solid $reflect-input-focus-color;
    }

    &::placeholder {
      color: lighten($reflect-input-color, 50%);
    }
  }

  textarea {
    resize: none;
  }
}
