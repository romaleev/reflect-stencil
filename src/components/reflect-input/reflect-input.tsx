import { Component, Event, Method, Host, State, h, EventEmitter } from '@stencil/core'
import { Regex, InputData } from '../../utils/utils'

const REGEX_NAME = Regex.name.toString().slice(1, -1)
const REGEX_EMAIL = Regex.email.toString().slice(1, -1)

@Component({
  tag: 'reflect-input',
  styleUrl: 'reflect-input.scss',
  shadow: true
})

/**
 * ReflectInput contains input fields which are accessible as a single, combined value
 *
 * Emits `inputChanged` event per input value change
 *
 * 'getData(): [[InputData]]' get inputs data
 * 'setData(data: [[InputData]]): void' set inputs data
 * 'clear(): void' - clear inputs data
 *
 * ```
 * <reflect-input/>
 * ```
 */
export class ReflectInput {

  @State() name = ''
  @State() email = ''
  @State() message = ''

  private getCombined(): InputData { // returns combined value
    return { name: this.name, email: this.email, message: this.message }
  }

  private onInput = (event) => { // input callback
    this[event.target.name] = event.target.value
    this.inputChanged.emit(this.getCombined())
  }

  /**
   * @event `inputChanged` emitted per input value change
   * @public
   */
  @Event({
    eventName: 'inputChanged',
    composed: true,
    cancelable: true,
    bubbles: true,
  }) inputChanged: EventEmitter<InputData>

  /**
   * @returns data with input fields values of type [[InputData]].
   * @public
   */
  @Method()
  async getData():Promise<InputData> {
    return this.getCombined()
  }

  /**
   * @param data with input fields values of type [[InputData]].
   * @public
   */
  @Method()
  async setData(data: InputData):Promise<void> {
    Object.keys(data).forEach((field: string) =>
      this[field] = data[field])
    this.inputChanged.emit(this.getCombined())
  }

  /**
   * Clears input contents
   * @public
   */
  @Method()
  async clear() {
    await this.setData({ name: '', email: '', message: '' })
  }

  render() {
    return (
      <Host>
        <input name='name'
               placeholder="NAME*"
               type="text"
               value={this.name}
               onInput={this.onInput}
               tabindex="1"
               pattern={REGEX_NAME}
               autoFocus
               maxlength="30"
               required/>
        <input name='email'
               type="email"
               placeholder="EMAIL*"
               value={this.email}
               onInput={this.onInput}
               tabindex="2"
               pattern={REGEX_EMAIL}
               maxlength="30"
               required/>
        <textarea name='message'
                  placeholder="MESSAGE*"
                  value={this.message}
                  onInput={this.onInput}
                  tabindex="3"
                  rows={3}
                  maxlength="500"
                  required/>
      </Host>
    )
  }

}
