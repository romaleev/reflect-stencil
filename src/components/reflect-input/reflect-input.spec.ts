import { newSpecPage } from '@stencil/core/testing'
import { ReflectInput } from './reflect-input'

const MockEmptyInput = { name: '', email: '', message: ''}
const MockFilledInput = { name: 'Roma', email: 'email@example.com', message: 'Hi'}

describe('reflect-input', () => {
  it('builds and renders', () => {
    const reflectInput = new ReflectInput()
    expect(reflectInput).toBeTruthy()
    expect(reflectInput).toMatchSnapshot();
  })

  it('should have getData(), setData(...), clear() methods and inputChanged event working property', async () => {

    const page = await newSpecPage({
      components: [ ReflectInput ],
      html: `<reflect-input></reflect-input>`
    })

    let eventSpy = jest.fn()
    page.doc.addEventListener('inputChanged', eventSpy)

    const reflectInput = await document.querySelector('reflect-input')

    expect(await reflectInput.getData()).toEqual(MockEmptyInput)

    await reflectInput.setData(MockFilledInput)

    expect(eventSpy).toHaveBeenCalled()
    expect(eventSpy.mock.calls[eventSpy.mock.calls.length - 1][0].detail).toEqual(MockFilledInput)

    expect(await reflectInput.getData()).toEqual(MockFilledInput)

    await reflectInput.clear()

    expect(eventSpy).toHaveBeenCalled()
    expect(eventSpy.mock.calls[eventSpy.mock.calls.length - 1][0].detail).toEqual(MockEmptyInput)

    expect(await reflectInput.getData()).toEqual(MockEmptyInput)
  })
})
