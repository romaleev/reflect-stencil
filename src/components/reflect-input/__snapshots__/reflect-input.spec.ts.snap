// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reflect-input builds and renders 1`] = `
ReflectInput {
  "email": "",
  "inputChanged": Object {
    "emit": [Function],
  },
  "message": "",
  "name": "",
  "onInput": [Function],
}
`;
